jax/scipy/sparse/test_gmres.py

```diff
@@ -0,0 +1,152 @@
+  Q = np.zeros((n, n_kry + 1), dtype=x0.dtype)
+  Q[:, 0] = x0/jax.numpy.linalg.norm(x0)
+  Q = jnp.array(Q)
+  H = jax.numpy.zeros((n_kry, n_kry + 1), dtype=x0.dtype)
+  tol = A.size*jax.numpy.finfo(dtype).eps
+  M = linalg._identity
+
+  @jax.tree_util.Partial
+  def A_mv(x):
+    return A @ x
+  for k in range(n_kry):
+    Q, H = linalg.kth_arnoldi_iteration(k, A_mv, M, Q, H, tol)
+  QAQ = Q[:, :n_kry].conj().T @ A @ Q[:, :n_kry]
+  try:
+    np.testing.assert_allclose(H.T[:n_kry, :], QAQ, atol=tol)
+  except AssertionError as err:
+    print(err)
+  else:
+    print(f"Arnoldi passed at dtype {dtype}!")
+
+if __name__ == "__main__":
+  for dtype in jax_dtypes:
+    test_gmres_on_small_known_problem(dtype)
+    test_gmres_arnoldi_step(dtype)
+
+
+#  @pytest.mark.parametrize("dtype", jax_dtypes)
+#  def test_gmres_krylov(dtype):
+#    """
+#    gmres_krylov correctly builds the QR-decomposed Arnoldi decomposition.
+#    This function assumes that gmres["kth_arnoldi_step (which is
+#    independently tested) is correct.
+#    """
+#    dummy = jax.numpy.zeros(1, dtype=dtype)
+#    dtype = dummy.dtype
+#    gmres = jitted_functions.gmres_wrapper(jax)
+
+#    n = 2
+#    n_kry = n
+#    np.random.seed(10)
+
+#    @jax.tree_util.Partial
+#    def A_mv(x):
+#      return A @ x
+#    A = jax.numpy.array(np.random.rand(n, n).astype(dtype))
+#    tol = A.size*jax.numpy.finfo(dtype).eps
+#    x0 = jax.numpy.array(np.random.rand(n).astype(dtype))
+#    b = jax.numpy.array(np.random.rand(n), dtype=dtype)
+#    r, beta = gmres.gmres_residual(A_mv, [], b, x0)
+#    _, V, R, _ = gmres.gmres_krylov(A_mv, [], n_kry, x0, r, beta,
+#                                    tol, jax.numpy.linalg.norm(b))
+#    phases = jax.numpy.sign(jax.numpy.diagonal(R[:-1, :]))
+#    R = phases.conj()[:, None] * R[:-1, :]
+#    Vtest = np.zeros((n, n_kry + 1), dtype=x0.dtype)
+#    Vtest[:, 0] = r/beta
+#    Vtest = jax.numpy.array(Vtest)
+#    Htest = jax.numpy.zeros((n_kry + 1, n_kry), dtype=x0.dtype)
+#    for k in range(n_kry):
+#      Vtest, Htest = gmres.kth_arnoldi_step(k, A_mv, [], Vtest, Htest, tol)
+#    _, Rtest = jax.numpy.linalg.qr(Htest)
+#    phases = jax.numpy.sign(jax.numpy.diagonal(Rtest))
+#    Rtest = phases.conj()[:, None] * Rtest
+#    np.testing.assert_allclose(V, Vtest, atol=tol)
+#    np.testing.assert_allclose(R, Rtest, atol=tol)
+
+
+
+
+#  @pytest.mark.parametrize("dtype", jax_dtypes)
+#  def test_givens(dtype):
+#    """
+#    gmres["givens_rotation produces the correct rotation factors.
+#    """
+#    gmres = jitted_functions.gmres_wrapper(jax)
+#    np.random.seed(10)
+#    v = jax.numpy.array(np.random.rand(2).astype(dtype))
+#    cs, sn = gmres.givens_rotation(*v)
+#    rot = np.zeros((2, 2), dtype=dtype)
+#    rot[0, 0] = cs
+#    rot[1, 1] = cs
+#    rot[0, 1] = -sn
+#    rot[1, 0] = sn
+#    rot = jax.numpy.array(rot)
+#    result = rot @ v
+#    tol = 4*jax.numpy.finfo(dtype).eps
+#    np.testing.assert_allclose(result[-1], 0., atol=tol)
```

Comment:
One other unit test that would be nice to have is a consistency check with SciPy's GMRES after a small fixed number of iterations (before convergence). I have a version of that in my gist: https://gist.github.com/shoyer/dc33a5850337b6a87d48ed97b4727d29

Reply:
Looks great